### PR TITLE
[SFN] Add Support for MaxItem and MaxItemPath

### PR DESCRIPTION
**Pull request by MEPalma**
## Motivation
Currently, the SFN interpreter parses the MaxItem and MaxItemPath fields of ItemReader configuration productions, however these are not applied in the ItemReader configuration. Furthermore, the JsonPath extraction strategy for MaxItemPath fields did not perform valid value unpacking logic. This PR adds base behaviour for MaxItem and MaxItemPath fields, addresses the JsonPath unpacking issue, and adds relevant snapshot tests.

## Changes
- Implementation for MaxItem and MaxItemPath fields for both csv and json streams
- Fixed the preprocessing logic of MaxItemDecl
- Positive and negative MaxItem and MaxItemPath snapshot tests
- Fixes the JsonPath unpacking logic in MaxItemPath

### CODE REVIEW

File: localstack/services/stepfunctions/asl/component/state/state_execution/state_map/item_reader/reader_config/max_items_decl.py [hunk 1]
```diff
@@ -48,7 +65,45 @@ class MaxItemsPath(MaxItemsDecl):
...
+                        )
+                    ),
+                )
+            )
+
     def _get_value(self, env: Environment) -> int:
-        input_expr = parse(self.path)
-        max_items = input_expr.find(env.inp)
+        inp = env.stack[-1]
+        max_items = JSONPathUtils.extract_json(self.path, inp)
```
Comment (joe4dev) [inline localstack/services/stepfunctions/asl/component/state/state_execution/state_map/item_reader/reader_config/max_items_decl.py:107]: good to fix more occurrences of JSON extraction logic following https://github.com/localstack/localstack/pull/10521 👍

File: localstack/services/stepfunctions/asl/component/state/state_execution/state_map/item_reader/reader_config/reader_config_props.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+from typing import Any, Final
+
+from localstack.services.stepfunctions.asl.component.state.state_execution.state_map.item_reader.reader_config.max_items_decl import (
+    MaxItemsDecl,
+)
+from localstack.services.stepfunctions.asl.parse.typed_props import TypedProps
+
+
+class ReaderConfigProps(TypedProps):
+    _UNIQUE_SUB_INSTANCES: Final[set[type]] = {MaxItemsDecl}
```
Comment (joe4dev) [inline localstack/services/stepfunctions/asl/component/state/state_execution/state_map/item_reader/reader_config/reader_config_props.py:10]: that's meta with `type` being the type of the instance 🤯

File: localstack/services/stepfunctions/asl/component/state/state_execution/state_map/item_reader/resource_eval/resource_output_transformer/resource_output_transformer_json.py [hunk 3]
```diff
@@ -23,9 +23,7 @@
 
 class ResourceOutputTransformerJson(ResourceOutputTransformer):
     def _eval_body(self, env: Environment) -> None:
-        _: ReaderConfigOutput = (
-            env.stack.pop()
-        )  # Not used, but expected by workflow (hence should consume the stack).
+        reader_config: ReaderConfigOutput = env.stack.pop()
```
Comment (joe4dev) [inline localstack/services/stepfunctions/asl/component/state/state_execution/state_map/item_reader/resource_eval/resource_output_transformer/resource_output_transformer_json.py:26]: 🩹 🧹

File: tests/aws/services/stepfunctions/v2/scenarios/test_base_scenarios.py [hunk 4]
```diff
@@ -961,6 +961,84 @@ def test_map_item_reader_base_csv_headers_first_line(
...
+            create_iam_role_for_sfn,
+            create_state_machine,
+            sfn_snapshot,
+            definition,
+            exec_input,
+        )
+
+    @markers.aws.validated
+    @pytest.mark.parametrize(
+        "max_items_value", [-1, 0, 2, 1.5, 100_000_000, 100_000_001]
```
Comment (joe4dev) [inline tests/aws/services/stepfunctions/v2/scenarios/test_base_scenarios.py:1003]: is the 2 before 1.5 intentional?

File: localstack/services/stepfunctions/asl/component/state/state_execution/state_map/item_reader/reader_config/max_items_decl.py [hunk 5]
```diff
@@ -48,7 +65,45 @@ class MaxItemsPath(MaxItemsDecl):
     def __init__(self, path: str):
         self.path: Final[str] = path
 
+    def _validate_value(self, env: Environment, value: int) -> None:
+        if not isinstance(value, int):
+            # TODO: Note, this error appears to be validated at a earlier stage in AWS Step Functions, unlike the
```
Comment (joe4dev) [inline localstack/services/stepfunctions/asl/component/state/state_execution/state_map/item_reader/reader_config/max_items_decl.py:70]: interesting and worth commenting on 👍 . Just out of curiosity, how would we approach fixing this in the future?
* Does it require figuring out where it happens?
* Would we need to introduce extra validation steps?

**Review by joe4dev — APPROVED** [review]
LGTM 👍

File: tests/aws/services/stepfunctions/v2/scenarios/test_base_scenarios.py (hunk 4 above)
Comment (MEPalma) [reply tests/aws/services/stepfunctions/v2/scenarios/test_base_scenarios.py:1003]: I can definitely put them in order, this was a mild attempt to group valid inputs together but go lost during filtering the test inputs!

File: localstack/services/stepfunctions/asl/component/state/state_execution/state_map/item_reader/reader_config/max_items_decl.py (hunk 5 above)
Comment (MEPalma) [reply localstack/services/stepfunctions/asl/component/state/state_execution/state_map/item_reader/reader_config/max_items_decl.py:70]: This needs more validation tests first to accurately pin point when, and possibly why, the two validations are carried out at different depths. In terms of changes, it seems quite likely that the evaluation order within the map run will need some adjusting, ad some validation errors about MapItemPath are raised before MapStarted events, whilst some are raised after. This change would not be a trivial one, that is why I decided to postpone it to a future PR. As you identified, this PR mainly wanted to address the jsonpath extraction logic (and the max item evaluation as a result).

File: localstack/services/stepfunctions/asl/component/state/state_execution/state_map/item_reader/reader_config/reader_config_props.py (hunk 2 above)
Comment (MEPalma) [reply localstack/services/stepfunctions/asl/component/state/state_execution/state_map/item_reader/reader_config/reader_config_props.py:10]: Indeed, ASL Component and EvaluationComponents are stored and retrieved via type through TypedProps objects (here is an example https://github.com/localstack/localstack/blob/master/localstack/services/stepfunctions/asl/parse/preprocessor.py#L592). This is a strategy to detect duplicate or conflicting language productions during preprocessing.

File: localstack/services/stepfunctions/asl/component/state/state_execution/state_map/item_reader/reader_config/max_items_decl.py (hunk 5 above)
Comment (joe4dev) [reply localstack/services/stepfunctions/asl/component/state/state_execution/state_map/item_reader/reader_config/max_items_decl.py:70]: Very reasonable 👍 Thanks for the clarification.

**Pull request merged**